internal/ioutil/io.go

```diff
@@ -36,16 +36,19 @@ func (fn CloserFunc) Close() error {
 // The read content is verified against the size and the digest.
 func ReadAll(r io.Reader, desc ocispec.Descriptor) ([]byte, error) {
 	// verify while reading
-	verifier := desc.Digest.Verifier()
-	r = io.TeeReader(r, verifier)
+	if desc.Digest != "" {
+		verifier := desc.Digest.Verifier()
+		r = io.TeeReader(r, verifier)
+		if !verifier.Verified() {
```

Comment:
I think `verifier.Verified()` should not be called before `io.ReadFull(r, buf);` is done. After `ReadFull` is done, the content stream is read (in the direction of below diagram) and digest can be properly calcaulated. cc @Wwwsylvia Can you help confirm?
```
            (r)                    (r)
+-----+ input reader +-----> output reader +-------->
                                +
                                |
                                +
                             verifier +----------->
```

Reply:
Thanks for taking a look @qweeah; after going through this with @michaelb990 today, we realized there is actually another problem here.  At no point are we calculating the digest in `oras-go`, but instead, trusting what the server happens to return.  In the general case, that's probably fine, but I think the better approach is to use the digest returned by the server only for additional validation, as opposed to using it as the source-of-truth.

I believe what you're referring to is in reference to `TeeReader` which imposes a restriction of the read order, is that correct?

Reply:
Yes. The `TeeReader` requires reader `r` to be fully read so the content can be written to the writer `verifier`.